### ISSUE TITLE
Support #line

### ISSUE THREAD

**Issue by chrysn** [doxygen/doxygen#10697]
**Describe the bug**

When C code is generated as an input to Doxygen (eg. because it is the output of preprocessing that Doxygen can not do, or because it documents defines that do not originate in source code but are provided), those file names and line numbers in that file will not be useful. However, there *are* file names and line numbers useful to the reader -- and the generator can annotate them using the same [preprocessor line control directive (`#line`)](https://learn.microsoft.com/en-us/cpp/preprocessor/hash-line-directive-c-cpp) generated by compilers such as GCC as part of their C preprocessing.

For example, it may generate a file `from-make.h` as follows:

```c
/* This file was generated by myscript.py */
/** @file */
#line 1234 "makefiles/cflags.mk"
/** @brief Configured Foo value */
#define FOO 1234
#line 987 "makefiles/configuration.mk"
/** @brief The current Fjord's Narf */
#define NARF NARF_THIS
```

**Expected behavior**

When Doxygen generates the help for FOO, it should say "Definition at line 1235 of file makefiles/cflags.mk". Instead, it says "Definition at line 5 of from-make.h", which is ignoring the directive, and not helpful to the user (who probably does not have that file because it is only generated for Doxygen, and even if they had, it is not where the definition originally comes from).

If the indicated source file is not part of Doxygen's source browser, adding it once mentioned or stating the location without having a link would work, although I'd prefer the former. (If it helps, the source files may be marked explicitly for inclusion in the source browser).

**To Reproduce**

Save the snippet above as from-make.h, and run Doxygen with defaults from `-g`, just setting `SOURCE_BROWSER = YES`

**Version**

Tested with Doxygen 1.9.8 as packaged for Debian GNU/Linux trixie/sid.

**Comment by albert-github** [comment]
Interesting idea.
You wrote:
>who probably does not have that file because it is only generated for Doxygen, and even if they had, it is not where the definition originally comes from

and when looking through the given link we see:
![image](https://github.com/doxygen/doxygen/assets/5223533/79647dc5-20b5-498e-b261-1ed42f19aba9)

so he will be pointed in the right direction.

When you would have a link like: "Definition at line 1235 of file makefiles/cflags.mk" this would mean that doxygen should have the accessibility to the mentioned file and display the source code including some links to the right lines.

An alternative might be that the line would read something like:
"Definition at line 5 of from-make.h generated from line 1235 of file makefiles/cflags.mk".
where the parts "5" and "from-make.h" are hyperlinked and the rest isn't.
The disadvantage of this solution is that when the user has no access to the source code and only has access to the documentation he still won't be able to see the original file.

**Comment by chrysn** [comment]
Right, they do get the data from there, but it's a needless redirection, and it makes an intermediate artifact visible.

Having the generated-from style text would be good improvement. At least if the files are part of the source tree (didn't yet check with which option to add them), they could still be linked. Personally, I'd prefer the file/line of the generated file to be hideable completly, but that's just a smaller matter then.

**Comment by albert-github** [comment]
>  Personally, I'd prefer the file/line of the generated file to be hideable completely, but that's just a smaller matter then.

Nice suggestion something line `\hidesourcedef`.

**Comment by AndreasMatthias** [comment]
> It would be great to have a command to tell doxygen the correct line number of each definition.

If we had such a command, then it would be great if this line number were also used in warning/error messages on the CLI.

```
/path/to/file.xxx:123: warning: Member foo of class XYZ is not documented.
```

**Comment by chrysn** [comment]
Processing in that LSP use case, I still do think that support for `#line` would be convenient, but as an alternative, if Doxygen had some generic and easy to produce input format (eg. plain text and commands to say that there is a "function" called this-and-that defined, either including file location data or with a @hidesourcedef and a manually created link), that would also satisfy these use cases.
